accord-core/src/main/java/accord/primitives/AbstractKeys.java

```diff
@@ -0,0 +1,346 @@
+            KeyRange range = rs.get(ri);
+            int nextai = range.nextHigherKeyIndex(this, ki + 1);
+            while (ki < nextai)
+            {
+                initialValue = fold.apply(ki, get(ki), param, initialValue);
+                if (initialValue == terminalValue)
+                    break done;
+                ++ki;
+            }
+        }
+
+        return initialValue;
+    }
+
+    /**
+     * A fold variation permitting more efficient operation over indices only, by providing ranges of matching indices
+     */
+    @Inline
+    public long rangeFoldl(KeyRanges rs, IndexedRangeFoldToLong fold, long param, long initialValue, long terminalValue)
+    {
+        int ki = 0, ri = 0;
+        while (true)
+        {
+            long rki = rs.findNextIntersection(ri, this, ki);
+            if (rki < 0)
+                break;
+
+            ri = (int)(rki >>> 32);
+            ki = (int)(rki);
+            KeyRange range = rs.get(ri);
+            int nextai = range.nextHigherKeyIndex(this, ki + 1);
+            initialValue = fold.apply(ki, nextai, param, initialValue);
+            if (initialValue == terminalValue)
+                break;
+            ki = nextai;
+        }
+
+        return initialValue;
+    }
+
+    @Inline
+    public long foldl(IndexedFoldToLong<? super K> fold, long param, long initialValue, long terminalValue)
+    {
+        for (int i = 0; i < keys.length; i++)
+        {
+            initialValue = fold.apply(i, keys[i], param, initialValue);
+            if (terminalValue == initialValue)
+                return initialValue;
+        }
+        return initialValue;
+    }
+
+    /**
+     * A fold variation that intersects two key sets, invoking the fold function only on those
+     * items that are members of both sets (with their corresponding indices).
+     */
+    @Inline
+    public long foldlIntersect(AbstractKeys<K, ?> intersect, IndexedFoldIntersectToLong<? super K> fold, long param, long initialValue, long terminalValue)
+    {
+        return SortedArrays.foldlIntersection(this.keys, intersect.keys, fold, param, initialValue, terminalValue);
+    }
+
+    /**
+     * A fold variation that invokes the fold function only on those items that are members of this set
+     * and NOT the provided set.
+     */
+    @Inline
+    public long foldlDifference(AbstractKeys<K, ?> subtract, IndexedFoldToLong<? super K> fold, long param, long initialValue, long terminalValue)
+    {
+        return SortedArrays.foldlDifference(keys, subtract.keys, fold, param, initialValue, terminalValue);
+    }
+
+    public Route toRoute(RoutingKey homeKey)
+    {
+        if (isEmpty())
+            return new Route(homeKey, new RoutingKey[] { homeKey });
+
+        RoutingKey[] result = toRoutingKeysArray(homeKey);
+        int pos = Arrays.binarySearch(result, homeKey);
+        return new Route(result[pos], result);
+    }
+
+    public PartialRoute toPartialRoute(KeyRanges ranges, RoutingKey homeKey)
```

Comment:
Let's leave cleaning this up to routables, if necessary

Regarding implementation, it makes sense to reference the one we already have rather than potentially reference two distinct homeKey?

Reply:
OK, what threw me off was `toRoutingKeysArray` which is actually copying the internal array and inserting homeKey if it isn't already there, and then we binary search to also reuse the existing homeKey copy.

And this seems to change again in Routables where I don't think this exists anymore because nothing that constructs a partial route works like